### PR TITLE
Assorted ASTVisitor work

### PR DESCRIPTION
**Pull request by sdkrystian**

### CODE REVIEW

File: include/mrdox/Metadata/Record.hpp [hunk 1]
```diff
@@ -67,12 +67,42 @@ struct BaseInfo
...
     std::vector<MemberRef> Records;
     std::vector<MemberRef> Functions;
     std::vector<MemberRef> Enums;
     std::vector<MemberRef> Types;
     std::vector<MemberRef> Fields;
     std::vector<MemberRef> Vars;
+
+// private:
+    template<typename T, typename... Args>
+    void emplace_back(Args&&... args)
```
Comment (vinniefalco) [inline include/mrdox/Metadata/Record.hpp:81]: No you can't put this here, because it is not a public API

File: include/mrdox/Metadata/Record.hpp [hunk 2]
```diff
@@ -67,12 +67,42 @@ struct BaseInfo
...
+            else if constexpr(std::is_same_v<U, TypedefInfo>)
+                Types.emplace_back(std::forward<Args>(args)...);
+            else if constexpr(std::is_same_v<U, EnumInfo>)
+                Enums.emplace_back(std::forward<Args>(args)...);
+            else if constexpr(std::is_same_v<U, FieldInfo>)
+                Fields.emplace_back(std::forward<Args>(args)...);
+            else if constexpr(std::is_same_v<U, VarInfo>)
+                Vars.emplace_back(std::forward<Args>(args)...);
+            else
+                llvm_unreachable("invalid RecordScope member");
```
Comment (vinniefalco) [inline include/mrdox/Metadata/Record.hpp:99]: we need to REMOVE calls to llvm in public interfaces not ADD them.

File: include/mrdox/Metadata/Scope.hpp [hunk 3]
```diff
@@ -47,6 +53,35 @@ struct Scope
         : isNamespaceScope(isNamespaceScope_)
     {
     }
+
+// private:
+    template<typename T, typename... Args>
+    void emplace_back(Args&&... args)
```
Comment (vinniefalco) [inline include/mrdox/Metadata/Scope.hpp:59]: Nope

File: source/-XML/XMLWriter.hpp [hunk 4]
```diff
@@ -85,7 +85,8 @@ class XMLWriter
     void writeSymbol(SymbolInfo const& I);
     void writeLocation(Location const& loc, bool def = false);
     void writeJavadoc(std::unique_ptr<Javadoc> const& javadoc);
-    void writeTemplate(const TemplateInfo& I);
+    void writeTemplateOpen(const std::unique_ptr<TemplateInfo>& I);
+    void writeTemplateClose(const std::unique_ptr<TemplateInfo>& I);
```
Comment (vinniefalco) [inline source/-XML/XMLWriter.hpp:89]: consider `openTemplate` and `closeTemplate`

File: source/AST/ASTVisitor.cpp [hunk 5]
```diff
@@ -32,7 +32,7 @@
 #include <llvm/Support/SHA1.h>
 #include <cassert>
 #include <ranges>
-
+#include <clang/Sema/Sema.h>
```
Comment (vinniefalco) [inline source/AST/ASTVisitor.cpp:35]: this should come before the std includes

File: source/AST/ASTVisitor.hpp [hunk 6]
```diff
@@ -210,48 +217,80 @@ class ASTVisitor
...
+        std::unique_ptr<TemplateInfo>&& = nullptr);
+    bool Traverse(CXXConstructorDecl*,
+        AccessSpecifier,
+        std::unique_ptr<TemplateInfo>&& = nullptr);
+    bool Traverse(CXXConversionDecl*,
+        AccessSpecifier,
+        std::unique_ptr<TemplateInfo>&& = nullptr);
+    bool Traverse(CXXDeductionGuideDecl*,
+        AccessSpecifier,
+        std::unique_ptr<TemplateInfo>&& = nullptr);
```
Comment (vinniefalco) [inline source/AST/ASTVisitor.hpp:267]: I like all this shit on one line even if it goes over

File: include/mrdox/Metadata/Scope.hpp [hunk 7]
```diff
@@ -13,9 +13,13 @@
 #define MRDOX_METADATA_SCOPE_HPP
 
 #include <mrdox/Platform.hpp>
+#include <mrdox/Error.hpp>
+#include <mrdox/MetadataFwd.hpp>
 #include <mrdox/Metadata/Reference.hpp>
 #include <mrdox/Metadata/Typedef.hpp>
 #include <mrdox/Metadata/Enum.hpp>
+#include "Support/Debug.hpp"
```
Comment (vinniefalco) [inline include/mrdox/Metadata/Scope.hpp:21]: No you can't include this file, it is not public

File: include/mrdox/Metadata/Record.hpp (hunk 2 above)
Comment (sdkrystian) [reply include/mrdox/Metadata/Record.hpp:99]: I was going to ask to add an `Unreachable` macro to `Platform.hpp`... so maybe I shall do that

File: include/mrdox/Metadata/Access.hpp [hunk 8]
```diff
@@ -44,6 +44,16 @@ struct MemberRef
 {
     SymbolID id;
     Access access;
+
+    constexpr MemberRef() = default;
+
+    constexpr MemberRef(
+        const SymbolID& id,
```
Comment (vinniefalco) [inline include/mrdox/Metadata/Access.hpp:51]: please don't use the same formal parameter names as data members, change this to `id_`

File: source/-XML/XMLWriter.hpp [hunk 9]
```diff
@@ -85,7 +85,8 @@ class XMLWriter
     void writeSymbol(SymbolInfo const& I);
     void writeLocation(Location const& loc, bool def = false);
     void writeJavadoc(std::unique_ptr<Javadoc> const& javadoc);
-    void writeTemplate(const TemplateInfo& I);
+    void openTemplate(const std::unique_ptr<TemplateInfo>& I);
+    void closeTemplate(const std::unique_ptr<TemplateInfo>& I);
```
Comment (vinniefalco) [inline source/-XML/XMLWriter.hpp:89]: We need to think about standardizing on east const...

Comment (sdkrystian) [reply source/-XML/XMLWriter.hpp:89]: maaaan

File: source/AST/ASTVisitor.cpp [hunk 10]
```diff
@@ -748,10 +591,71 @@ extractInfo(
...
+{
+    const DeclContext* DC = D->getDeclContext();
+    do
+    {
+        // KRYSTIAN TODO: this can be a switch using DeclContext::getDeclKind
+        if(const auto* N = dyn_cast<NamespaceDecl>(DC))
+        {
+
+            Namespaces.emplace_back(
+                extractSymbolID(N));
```
Comment (vinniefalco) [inline source/AST/ASTVisitor.cpp:610]: you didn't need to wrap this line but w/e

File: source/AST/ASTVisitor.cpp [hunk 11]
```diff
@@ -748,10 +591,71 @@ extractInfo(
...
+            const CXXRecordDecl* RD = N->getTemplateInstantiationPattern();
+#if 0
+            std::vector<TArg> args;
+            buildTemplateArgs(args, N->getTemplateArgs().asArray());
+            SymbolID primary = extractSymbolID(RD);
+#endif
+
+            Namespaces.emplace_back(
+                // extractSymbolID(N));
+                extractSymbolID(RD));
```
Comment (vinniefalco) [inline source/AST/ASTVisitor.cpp:626]: there's no need to wrap this when you have longer lines surrounding it

File: source/AST/ASTVisitorHelpers.hpp [hunk 12]
```diff
@@ -0,0 +1,79 @@
+//
+// Copyright (c) 2023 Krystian Stasiowski (sdkrystian@gmail.com)
+//
+// Official repository: https://github.com/cppalliance/mrdox
```
Comment (vinniefalco) [inline source/AST/ASTVisitorHelpers.hpp:4]: Missing the license

File: source/AST/ASTVisitorHelpers.hpp [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+//
+
+#ifndef MRDOX_API_AST_ASTVISITORHELPERS_HPP
+#define MRDOX_API_AST_ASTVISITORHELPERS_HPP
+
+#include <mrdox/Platform.hpp>
+#include <mrdox/MetadataFwd.hpp>
+#include <mrdox/Metadata/Record.hpp>
+#include <mrdox/Metadata/Scope.hpp>
+#include "Support/Debug.hpp"
```
Comment (vinniefalco) [inline source/AST/ASTVisitorHelpers.hpp:14]: This is out of order, private includes come first

File: source/AST/ASTVisitorHelpers.hpp [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+            llvm_unreachable("invalid Scope child");
+    }
+    else
+    {
+        llvm_unreachable("invalid arguments");
+    }
+}
+
+template<typename T, typename... Args>
+void insertChild(RecordScope& P, Args&&... args)
```
Comment (vinniefalco) [inline source/AST/ASTVisitorHelpers.hpp:50]: This looks pretty silly now, when it should just be a set of overloaded inline free functions :)

I don't care for now though.

Comment (sdkrystian) [reply source/AST/ASTVisitorHelpers.hpp:50]: a bit overkill, yeah

**Pull request merged**